File: app/routers/recurring.py
from datetime import datetime, timedelta
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth import TokenData, get_current_user
from app.database import get_db
from app.models import RecurringTransaction, RecurringFrequency

router = APIRouter(prefix="/api/v1/recurring", tags=["recurring"])


class CreateRecurringRequest(BaseModel):
    merchant: str
    amount: float
    currency: str = "USD"
    frequency: str
    category_id: Optional[int] = None
    sub_category_id: Optional[int] = None
    next_due_date: str
    notes: str = ""


class UpdateRecurringRequest(BaseModel):
    merchant: Optional[str] = None
    amount: Optional[float] = None
    frequency: Optional[str] = None
    next_due_date: Optional[str] = None
    is_active: Optional[bool] = None
    notes: Optional[str] = None


def _rec_dict(r: RecurringTransaction) -> dict:
    return {
        "id": r.id, "family_id": r.family_id, "merchant": r.merchant,
        "amount": r.amount, "currency": r.currency, "frequency": r.frequency.value,
        "category_id": r.category_id, "sub_category_id": r.sub_category_id,
        "next_due_date": str(r.next_due_date), "is_active": r.is_active,
        "auto_detected": r.auto_detected, "notes": r.notes,
        "created_at": str(r.created_at), "updated_at": str(r.updated_at),
    }


@router.get("")
async def list_recurring(
    current_user: TokenData = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    rows = (await db.execute(
        select(RecurringTransaction).where(RecurringTransaction.family_id == current_user.family_id)
        .order_by(RecurringTransaction.next_due_date)
    )).scalars().all()
    return [_rec_dict(r) for r in rows]


@router.post("", status_code=status.HTTP_201_CREATED)
async def create_recurring(
    req: CreateRecurringRequest,
    current_user: TokenData = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    r = RecurringTransaction(
        family_id=current_user.family_id, created_by_user_id=current_user.user_id,
        merchant=req.merchant, amount=req.amount, currency=req.currency,
        frequency=RecurringFrequency(req.frequency),
        category_id=req.category_id, sub_category_id=req.sub_category_id,
        next_due_date=datetime.fromisoformat(req.next_due_date),
        is_active=True, notes=req.notes,
    )
    db.add(r)
    await db.commit()
    await db.refresh(r)
    return _rec_dict(r)


@router.get("/upcoming")
async def get_upcoming(
    current_user: TokenData = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    cutoff = datetime.utcnow() + timedelta(days=30)
    rows = (await db.execute(
        select(RecurringTransaction).where(
            RecurringTransaction.family_id == current_user.family_id,
            RecurringTransaction.is_active == True,
            RecurringTransaction.next_due_date <= cutoff,
        ).order_by(RecurringTransaction.next_due_date)
    )).scalars().all()
    return [_rec_dict(r) for r in rows]


@router.put("/{id}")
async def update_recurring(
    id: int, req: UpdateRecurringRequest,
    current_user: TokenData = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    r = (await db.execute(
        select(RecurringTransaction).where(
            RecurringTransaction.id == id, RecurringTransaction.family_id == current_user.family_id)
    )).scalar_one_or_none()
    if not r:
        raise HTTPException(404, "Not found")
    if req.merchant is not None: r.merchant = req.merchant
    if req.amount is not None: r.amount = req.amount
    if req.frequency is not None: r.frequency = RecurringFrequency(req.frequency)
    if req.next_due_date is not None: r.next_due_date = datetime.fromisoformat(req.next_due_date)
    if req.is_active is not None: r.is_active = req.is_active
    if req.notes is not None: r.notes = req.notes
    await db.commit()
    await db.refresh(r)
    return _rec_dict(r)


@router.delete("/{id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_recurring(
    id: int,
    current_user: TokenData = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    r = (await db.execute(
        select(RecurringTransaction).where(
            RecurringTransaction.id == id, RecurringTransaction.family_id == current_user.family_id)
    )).scalar_one_or_none()
    if not r:
        raise HTTPException(404, "Not found")
    await db.delete(r)
    await db.commit()

File: handlers/recurring.go
package handlers

import (
	"net/http"
	"strconv"

	"wealthwatch/middleware"
	"wealthwatch/models"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"
)

type RecurringHandler struct {
	db *gorm.DB
}

func NewRecurringHandler(db *gorm.DB) *RecurringHandler {
	return &RecurringHandler{db: db}
}

type CreateRecurringRequest struct {
	Merchant      string                    `json:"merchant" binding:"required"`
	Amount        float64                   `json:"amount" binding:"required,gt=0"`
	Currency      string                    `json:"currency"`
	Frequency     models.RecurringFrequency `json:"frequency" binding:"required"`
	CategoryID    *uint                     `json:"category_id"`
	SubCategoryID *uint                     `json:"sub_category_id"`
	NextDueDate   string                    `json:"next_due_date" binding:"required"`
	Notes         string                    `json:"notes"`
}

func (h *RecurringHandler) CreateRecurring(c *gin.Context) {
	userID, ok := middleware.GetUserID(c)
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User not authenticated"})
		return
	}
	familyID, ok := middleware.GetFamilyID(c)
	if !ok {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Family not found in token"})
		return
	}

	var req CreateRecurringRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	nextDue, err := parseDate(req.NextDueDate)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid next_due_date format"})
		return
	}

	currency := req.Currency
	if currency == "" {
		currency = "USD"
	}

	rec := models.RecurringTransaction{
		FamilyID:        familyID,
		CreatedByUserID: userID,
		Merchant:        req.Merchant,
		Amount:          req.Amount,
		Currency:        currency,
		Frequency:       req.Frequency,
		CategoryID:      req.CategoryID,
		SubCategoryID:   req.SubCategoryID,
		NextDueDate:     nextDue,
		IsActive:        true,
		AutoDetected:    false,
		Notes:           req.Notes,
	}

	if err := h.db.Create(&rec).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create recurring transaction"})
		return
	}

	h.db.Preload("Category").Preload("SubCategory").First(&rec, rec.ID)
	c.JSON(http.StatusCreated, rec)
}

func (h *RecurringHandler) ListRecurring(c *gin.Context) {
	familyID, ok := middleware.GetFamilyID(c)
	if !ok {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Family not found in token"})
		return
	}

	var items []models.RecurringTransaction
	q := h.db.Preload("Category").Preload("SubCategory").Where("family_id = ?", familyID).Order("next_due_date ASC")

	if active := c.Query("active"); active != "" {
		q = q.Where("is_active = ?", active == "true")
	}

	if err := q.Find(&items).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch recurring transactions"})
		return
	}

	c.JSON(http.StatusOK, items)
}

func (h *RecurringHandler) UpdateRecurring(c *gin.Context) {
	familyID, ok := middleware.GetFamilyID(c)
	if !ok {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Family not found in token"})
		return
	}

	id, err := strconv.ParseUint(c.Param("id"), 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid ID"})
		return
	}

	var rec models.RecurringTransaction
	if err := h.db.Where("id = ? AND family_id = ?", uint(id), familyID).First(&rec).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Recurring transaction not found"})
		return
	}

	var body map[string]interface{}
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if err := h.db.Model(&rec).Updates(body).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update"})
		return
	}

	h.db.Preload("Category").Preload("SubCategory").First(&rec, rec.ID)
	c.JSON(http.StatusOK, rec)
}

func (h *RecurringHandler) DeleteRecurring(c *gin.Context) {
	familyID, ok := middleware.GetFamilyID(c)
	if !ok {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Family not found in token"})
		return
	}

	id, err := strconv.ParseUint(c.Param("id"), 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid ID"})
		return
	}

	var rec models.RecurringTransaction
	if err := h.db.Where("id = ? AND family_id = ?", uint(id), familyID).First(&rec).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Not found"})
		return
	}

	if err := h.db.Delete(&rec).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete"})
		return
	}

	c.Status(http.StatusNoContent)
}

func (h *RecurringHandler) GetUpcoming(c *gin.Context) {
	familyID, ok := middleware.GetFamilyID(c)
	if !ok {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Family not found in token"})
		return
	}

	var items []models.RecurringTransaction
	if err := h.db.Preload("Category").Preload("SubCategory").
		Where("family_id = ? AND is_active = ? AND next_due_date <= NOW() + INTERVAL '30 days'", familyID, true).
		Order("next_due_date ASC").
		Find(&items).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch upcoming"})
		return
	}

	c.JSON(http.StatusOK, items)
}
